### ISSUE TITLE
Variables not resolved with include statements

### ISSUE THREAD

**Issue by muktip** [ansible/ansible#16990]
##### ISSUE TYPE

- Bug Report
##### ANSIBLE VERSION

```
ansible 2.1.1.0
```
##### CONFIGURATION

None
##### OS / ENVIRONMENT

N/A
##### SUMMARY

Variables do not resolve and error out as undefined when using _include_ statements in tasks. When you replace the below code for /duck/main.yml with anything but 'include' statements, it works fine. I have verified that this is not an issue in 2.1.0.
##### STEPS TO REPRODUCE

tasks/main.yml

```

---

  - set_fact:
      folder_name: 'duck'

  - debug:
      var: folder_name

  - include: "{{ folder_name }}/main.yml"

```

tasks/duck/main.yml

```

---
  - include: test.yml
```

tasks/duck/test.yml

```

---
  - debug:
      msg: "I made it here!"
```
##### EXPECTED RESULTS

```
PLAY [Set Facts Issue] *********************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [foo : set_fact] **********************************************************
task path: ansible/roles/foo/tasks/main.yml:3
ok: [localhost] => {"ansible_facts": {"folder_name": "duck"}, "changed": false, "invocation": {"module_args": {"folder_name": "duck"}, "module_name": "set_fact"}}

TASK [foo : debug] *************************************************************
task path: ansible/roles/foo/tasks/main.yml:6
ok: [localhost] => {
    "folder_name": "duck"
}

TASK [foo : include] ***********************************************************
task path: ansible/roles/foo/tasks/main.yml:8
included: ansible/roles/foo/tasks/duck/main.yml for localhost

TASK [foo : include] ***********************************************************
task path: ansible/roles/foo/tasks/duck/main.yml:2
included: ansible/roles/foo/tasks/duck/test.yml for localhost

TASK [foo : debug] *************************************************************
task path: ansible/roles/foo/tasks/duck/test.yml:2
ok: [localhost] => {
    "msg": "I made it here!"
}

PLAY RECAP *********************************************************************
localhost                  : ok=6    changed=0    unreachable=0    failed=0
```
##### ACTUAL RESULTS

```
PLAY [Set Facts Issue] *********************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [foo : set_fact] **********************************************************
task path: ansible/roles/foo/tasks/main.yml:3
ok: [localhost] => {"ansible_facts": {"folder_name": "duck"}, "changed": false, "invocation": {"module_args": {"folder_name": "duck"}, "module_name": "set_fact"}}

TASK [foo : debug] *************************************************************
task path: ansible/roles/foo/tasks/main.yml:6
ok: [localhost] => {
    "folder_name": "duck"
}

TASK [foo : include] ***********************************************************
task path: ansible/roles/foo/tasks/main.yml:8
fatal: [localhost]: FAILED! => {"failed": true, "reason": "'folder_name' is undefined"}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=1
```

**Comment by jimi-c** [comment]
This appears to be (mostly) fixed on at least devel, with one caveat: the code loading the includes try to do so as "static" includes first. In the above example, I happened to name my top-level playbook `test.yml`, so when the static include occurred, it found that playbook and raised an error:

<pre>
TASK [test : include] **********************************************************
task path: /home/jamesc/testing/16990/roles/test/tasks/main.yml:7
fatal: [localhost]: FAILED! => {"failed": true, "reason": "no action detected in task. This often indicates a misspelled module name, or incorrect module path.\n\nThe error appears to have been in '/home/jamesc/testing/16990/test.yml': line 1, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- hosts: localhost\n  ^ here\n\n\nThe error appears to have been in '/home/jamesc/testing/16990/test.yml': line 1, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- hosts: localhost\n  ^ here\n"}
</pre>

Renaming the include in the duck directory to `test_duck.yml` fixed this:

<pre>
[jamesc@jimi 16990]$ ansible-playbook -vv test.yml 
Using /etc/ansible/ansible.cfg as config file
PLAYBOOK: test.yml *************************************************************
1 plays in test.yml
PLAY [localhost] ***************************************************************
TASK [test : set_fact] *********************************************************
task path: /home/jamesc/testing/16990/roles/test/tasks/main.yml:3
ok: [localhost] => {"ansible_facts": {"folder_name": "duck"}, "changed": false}
TASK [test : debug] ************************************************************
task path: /home/jamesc/testing/16990/roles/test/tasks/main.yml:5
ok: [localhost] => {
    "folder_name": "duck"
}
TASK [test : include] **********************************************************
task path: /home/jamesc/testing/16990/roles/test/tasks/main.yml:7
[DEPRECATION WARNING]: Included file '/home/jamesc/testing/16990/test_duck.yml' not found, however since this include is not explicitly marked as 'static: yes', we will try and include it 
dynamically later. In the future, this will be an error unless 'static: no' is used on the include task. If you do not want missing includes to be considered dynamic, use 'static: yes' on 
the include or set the global ansible.cfg options to make all inclues static for tasks and/or handlers.
This feature will be removed in a future release. Deprecation warnings can be disabled
 by setting deprecation_warnings=False in ansible.cfg.
included: /home/jamesc/testing/16990/roles/test/tasks/duck/main.yml for localhost
TASK [test : include] **********************************************************
task path: /home/jamesc/testing/16990/roles/test/tasks/duck/main.yml:1
included: /home/jamesc/testing/16990/roles/test/tasks/duck/test_duck.yml for localhost
TASK [test : debug] ************************************************************
task path: /home/jamesc/testing/16990/roles/test/tasks/duck/test_duck.yml:2
ok: [localhost] => {
    "msg": "I made it here!"
}
PLAY RECAP *********************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=0   
</pre>

You can see from the deprecated message that it tried to load something with a variable in its name statically, which is a bug.

**Comment by jimi-c** [comment]
# Closing This Ticket

Hi!

We believe the above commit should resolve this problem for you. This will also be included in the next release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!

**Comment by timw** [comment]
Since this is a regression from 2.1.0.0 -> 2.1.1.0, is there a plan to fix it in the 2.1.x series?

(At the moment there are so many breaks introduced in 2.1.1.0 that it's unusable, and we'll be dead-ended until 2.2 comes out).

**Comment by jimi-c** [comment]
We haven't really discussed a 2.1.2, we've been more focused on getting 2.2 ready. If that changes yes I would look to include this fix.

**Comment by timw** [comment]
It looks like 2.1.2 is on the plans now - can this be include?

**Comment by jimi-c** [comment]
@timw this was merged into stable-2.1 last night, and will be included in 2.1.2rc2 (due out today). Please test!
